### ISSUE TITLE
Mypy doesn't check for unbound variables

### ISSUE THREAD

**Issue by 1PaCHeK1** [python/mypy#16609]
**Bug Report**

Ignore `Unbound` variable

**To Reproduce**

Gist URL: https://gist.github.com/mypy-play/d2f10eb93875541fd255cd23f4c4a790
Playground URL: https://mypy-play.net/?mypy=latest&python=3.11&gist=d2f10eb93875541fd255cd23f4c4a790

```python
def func(flag: bool) -> str:
    if flag:
        name = "Name"
    return name


func(False)
```

**Expected Behavior**
error: unbound `name` variable

**Actual Behavior**
Type checking succeeds

- Mypy version used: 1.7.1
- Mypy command-line flags: None
- Mypy configuration options from `mypy.ini` (and other config files): Empty, this is reproducable on mypy playground
- Python version used: 3.11.3

**Comment by KotlinIsland** [comment]
Here is a more minified example:
```py
a: int
print(a)  # mypy does not report any error here
```

**Comment by hauntsaninja** [comment]
You can enable the possibly-undefined error code to catch this, see https://mypy-play.net/?mypy=latest&python=3.11&gist=4a516bee2f5f8614fbc76f01e3d8c483
